### PR TITLE
Add tectonic build script.

### PR DESCRIPTION
**Pull request by MichaelHatherly**
Seems to be building fine now after @giordano's ICU and harfbuzz fixes. (I built locally on linux, win, and mac but not the others, hopefully there aren't too many problems with those.)

### CODE REVIEW

**Comment by giordano** [comment]
Doesn't seem to like 32-bit systems and musl?

File: T/tectonic/build_tarballs.jl [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+# Bash recipe for building across all platforms
+script = raw"""
+cd $WORKSPACE/srcdir/tectonic-*/
+cargo build --release
+if [[ ${target} == *mingw* ]]; then
+    cp target/${rust_target}/release/tectonic.exe $bindir/
+else
+    cp target/${rust_target}/release/tectonic $bindir/
+fi
```
Comment (giordano) [inline T/tectonic/build_tarballs.jl:24]: [We have you covered](https://juliapackaging.github.io/BinaryBuilder.jl/dev/build_tips/#Automatic-environment-variables):
```suggestion
cp target/${rust_target}/release/tectonic${exeext} ${bindir}/
```

File: T/tectonic/build_tarballs.jl [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    ArchiveSource(
+        "https://github.com/tectonic-typesetting/tectonic/archive/tectonic@$(version).tar.gz",
+        "0e55188eafc1b58f3660a303fcdd6adc071051b9eb728119837fbeed2309914f"
+    )
+]
+
+# Bash recipe for building across all platforms
+script = raw"""
+cd $WORKSPACE/srcdir/tectonic-*/
+cargo build --release
```
Comment (giordano) [inline T/tectonic/build_tarballs.jl:19]: This should speedup compilation a bit:
```suggestion
cargo build --release -j${nproc}
```

File: T/tectonic/build_tarballs.jl [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    Dependency("Graphite2_jll"),
+    Dependency("HarfBuzz_jll"),
+    Dependency("ICU_jll"),
+    Dependency("OpenSSL_jll"),
+    Dependency("Zlib_jll"),
+    Dependency("libpng_jll"),
+]
+
+# Build the tarballs, and possibly a `build.jl` as well.
+build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies; compilers=[:c, :rust], preferred_gcc_version=v"7")
```
Comment (giordano) [inline T/tectonic/build_tarballs.jl:50]:
```suggestion
build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies; compilers=[:c, :rust], preferred_gcc_version=v"7", lock_microarchitecture=false)
```

**Comment by giordano** [comment]
From a very quick search, I think the remaining platforms are failing because of https://github.com/rust-lang/rust/issues/36710.  What do you want to do with them?

**Comment by MichaelHatherly** [comment]
Thanks for the edits.

> What do you want to do with them?

Not too bothered whether we support then or not currently, this is more platforms that are currently supported anyway so it's an improvement either way.

**Comment by giordano** [comment]
Ok, then I'd suggest to disable the failing platforms and I'll merge the pull request

**Pull request merged**